marlowe-dashboard-client/src/Dashboard/State.purs

```diff
@@ -292,6 +294,35 @@ handleAction input@{ currentSlot } (UpdateContract plutusAppId contractHistory@{
           $ for_ selectedStep' (handleAction input <<< ContractAction <<< Contract.MoveToStep)
       Nothing -> for_ (Contract.mkInitialState walletDetails currentSlot plutusAppId mempty contractHistory) (modifying _contracts <<< insert plutusAppId)
 
+-- This handler looks, in the given contract, for any payments to roles for which the current
+-- wallet holds the token, and then calls the "redeem" endpoint of the main marlowe app for each
+-- one, to make sure those funds reach the user's wallet (without the user having to do anything).
+-- The handler is called every time we receive a notification that the state of the contract's
+-- follower app has changed, so it will be called more often that is necessary. But there is no way
+-- to guard against that. (Well, we could keep track of payments redeemed in the application state,
+-- but that record would be lost when the browser is closed - and then those payments would all be
+-- redeemed again anyway the next time the user picks up the wallet. Since these duplicate requests
+-- will happen anyway, therefore, this extra complication doesn't seem worth the bother.
+handleAction input (RedeemPayments plutusAppId) = do
+  walletDetails <- use _walletDetails
+  contracts <- use _contracts
+  for_ (lookup plutusAppId contracts) \{ executionState, mMarloweParams, userParties } ->
+    for_ mMarloweParams \marloweParams ->
+      let
+        payments = getAllPayments executionState
+
+        isToParty party (Payment _ payee _) = case payee of
+          Party p -> p == party
+          _ -> false
+      in
+        for (List.fromFoldable userParties) \party ->
+          let
+            paymentsToParty = List.filter (isToParty party) payments
+          in
+            for paymentsToParty \payment -> case payment of
```

Comment:
4 levels of for seems a little nesty (pun intended), eventually I think runMaybeT could be of help, but no biggie.

Reply:
Fair point. I'll leave it for now, but revisit it later.